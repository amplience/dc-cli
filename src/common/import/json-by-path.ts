import axios from 'axios';
import { URL } from 'url';
import * as fs from 'fs';
import * as path from 'path';

export async function getJsonByPath(filename: string, relativeDir: string = __dirname): Promise<string> {
  if (filename.match(/^(http|https):\/\//)) {
    const result = await axios.get(filename);

    if (typeof result.data == 'string') {
      return result.data;
    }

    return JSON.stringify(result.data);
  }

  let resolvedFilename: string | URL = filename;
  if (filename.match(/file:\/\//)) {
    resolvedFilename = new URL(filename);
  } else if (filename.split(path.sep)[0].match(/^\.{1,2}$/)) {
    resolvedFilename = path.resolve(relativeDir, filename);
  }

  if (!fs.existsSync(resolvedFilename)) {
    throw new Error(
      `Cannot find JSON file "${filename}" using relative dir "${relativeDir}" (resolved path "${resolvedFilename}")`
    );
  }

  return fs.readFileSync(resolvedFilename, 'utf-8');
}
