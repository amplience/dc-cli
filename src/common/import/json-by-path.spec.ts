import axios from 'axios';
import fs from 'fs';
import path from 'path';
import { getJsonByPath } from './json-by-path';

jest.mock('axios');
jest.mock('fs');
jest.mock('path');

describe('content type schema helper', function() {
  beforeEach((): void => {
    jest.resetAllMocks();
  });

  async function successfulAxiosGetInvocation(protocol: string, schema: string, stringifyData: boolean): Promise<void> {
    const mockAxiosGet = axios.get as jest.Mock;
    const data = {
      $schema: 'test',
      id: 'test'
    };
    const mockSchemaResponse = {
      data: stringifyData ? JSON.stringify(data) : data
    };
    mockAxiosGet.mockResolvedValue(mockSchemaResponse);
    const response = await getJsonByPath(schema);
    expect(mockAxiosGet).toHaveBeenCalledWith(expect.stringMatching(protocol));
    expect(response).toEqual(JSON.stringify(data));
  }

  it('should load a schema from a url (http) as object', async function() {
    await successfulAxiosGetInvocation('http', 'http://example.com/schema.json', false);
  });

  it('should load a schema from a url (https) as object', async function() {
    await successfulAxiosGetInvocation('https', 'https://example.com/schema.json', false);
  });

  it('should load a schema from a url (http) as JSON string', async function() {
    await successfulAxiosGetInvocation('http', 'http://example.com/schema.json', true);
  });

  it('should load a schema from a url (https) as JSON string', async function() {
    await successfulAxiosGetInvocation('https', 'https://example.com/schema.json', true);
  });

  async function successfulLocalFileInvocation(schemaPath: string, relativePath: string = __dirname): Promise<void> {
    const mockFileRead = fs.readFileSync as jest.Mock;
    const mockExistsSync = fs.existsSync as jest.Mock;
    const mockSchemaData = {
      $schema: 'test',
      id: 'test'
    };
    mockFileRead.mockResolvedValue(JSON.stringify(mockSchemaData));
    mockExistsSync.mockResolvedValue(true);
    const response = await getJsonByPath(schemaPath, relativePath);
    expect(mockFileRead).toHaveBeenCalledTimes(1);
    expect(response).toEqual(JSON.stringify(mockSchemaData));
  }

  it('should load a schema from a local file (relative path) using __dirname as relative dir', async function() {
    const mockPathResolve = path.resolve as jest.Mock;
    await successfulLocalFileInvocation('./content-type-schema/schema.json');
    expect(mockPathResolve).toHaveBeenCalledWith(__dirname, './content-type-schema/schema.json');
  });

  it('should load a schema from a local file (relative path) using a supplied relative dir', async function() {
    const mockPathResolve = path.resolve as jest.Mock;
    await successfulLocalFileInvocation('./content-type-schema/schema.json', '/foo');
    expect(mockPathResolve).toHaveBeenCalledWith('/foo', './content-type-schema/schema.json');
  });

  it('should load a schema from a local file (with file url)', async function() {
    const mockPathResolve = path.resolve as jest.Mock;
    await successfulLocalFileInvocation('file://content-type-schema/schema.json');
    expect(mockPathResolve).not.toHaveBeenCalled();
  });

  it('should throw an error when file not found', async () => {
    const mockPathResolve = path.resolve as jest.Mock;
    const mockFileRead = fs.readFileSync as jest.Mock;
    const mockExistsSync = fs.existsSync as jest.Mock;
    mockExistsSync.mockReturnValue(false);

    await expect(getJsonByPath('./content-type-schema/schema.json', '/foo')).rejects.toThrowErrorMatchingSnapshot();

    expect(mockPathResolve).toHaveBeenCalledWith('/foo', './content-type-schema/schema.json');
    expect(mockExistsSync).toHaveBeenCalled();
    expect(mockFileRead).not.toHaveBeenCalled();
  });
});
