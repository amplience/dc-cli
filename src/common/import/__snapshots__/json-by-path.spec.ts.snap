// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content type schema helper should throw an error when file not found 1`] = `"Cannot find JSON file \\"./content-type-schema/schema.json\\" using relative dir \\"/foo\\" (resolved path \\"undefined\\")"`;
